File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第5章剪格子.cpp
//剪格子   lanqiaoOJ 211

#include<bits/stdc++.h>
using namespace std;
const int N=15;
int n, m;
int a[N][N], vis[N][N]; //格子是否被访问过
int sum, ans=100000;
int d[4][2] = { {1,0},{0,-1},{0,1},{-1,0} };  //4个方向
void dfs(int x, int y, int c, int s) {
    if (2*s>sum)  return;    //剪枝
    if (2*s==sum) {
        if (c<ans && vis[0][0]) ans = c; //左上角格子最少数量
        return;
    }
    vis[x][y]=1;
    for (int k=0; k<4; k++) {
        int tx=x+d[k][0], ty=y+d[k][1];
        if (tx>=0 && tx<n && ty>=0 && ty<m && !vis[tx][ty])
            dfs(tx,ty,c+1,s+a[x][y]);
    }
    vis[x][y]=0;
}
int main() {
    cin>>m>>n;
    for (int i=0; i<n; i++)
        for (int j=0; j<m; j++)
            cin>>a[i][j], sum+=a[i][j]; //求所有格子的和
    dfs(0,0,0,0);
    cout<<(ans==100000 ? 0 : ans);
    return 0;
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第5章剪格子.py
#剪格子   lanqiaoOJ 211

def dfs(x, y, c,s):
    global sum_num, ans
    if 2*s > sum_num: return
    if 2*s == sum_num:
        if ans>c and vis[0][0] == 1: ans = c
        return    
    vis[x][y] = 1
    dir = [(1, 0), (-1, 0), (0, -1), (0, 1)]
    for u,v in dir:
        tx, ty = x+u, y+v
        if tx >= 0 and tx <= n-1 and ty >= 0 and ty <= m-1:
            if vis[tx][ty] == 0:
                dfs(tx, ty, c+1, s+a[x][y])
    vis[x][y] = 0
    
m, n = map(int, input().split())
a = [list(map(int, input().split())) for _ in range(n)]
vis = [[0]*m for _ in range(n)]
sum_num = 0
for i in a:  sum_num += sum(i)  
if sum_num / 2 != sum_num // 2: print(0); exit() #无解
ans = 100000
dfs(0, 0, 0, 0)
print(ans)
